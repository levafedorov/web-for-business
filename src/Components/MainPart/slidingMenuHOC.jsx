import React, {Component} from 'react';

export default function slidingMenuHOC(Wrapped){

 class SlidingMenuHOC extends Component{
    
    state = {
        fixed: false
    }
    

    peelOffMenu = (ref) => {
            let scrolledSpace = window.scrollY || document.documentElement.scrollTop;
            this.resizeAlg(scrolledSpace);
    }
    

    resizeAlg = (scrolledSpace) => {
         
        if(scrolledSpace > 130 && !this.state.fixed){
            this.setState({fixed: true}); 
       }
       else if(scrolledSpace < 130 && this.state.fixed){
           this.setState({fixed: false}); 
       } 

    }


    componentDidMount(){
        window.addEventListener('scroll', this.peelOffMenu);
    }
      
    componentWillUnmount(){
        window.removeEventListener('scroll', this.peelOffMenu);
    }
 

    render(){
        const position = {
            top:0,
            position: this.state.fixed ? 'fixed' : 'relative'
        }      
  
        return <Wrapped position={position} {...this.props}/>
    }
} 

return SlidingMenuHOC 
}