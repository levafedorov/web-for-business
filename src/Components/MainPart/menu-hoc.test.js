import React from 'react';
import slidingMenuHOC from './SlidingMenuHoc';
import {shallow} from 'enzyme';

describe('HOCMenu component', () => {

    let MockedComponent; 
    let MenuHoc; 
    let didMountSpy; 
    let didUnmountSpy;
    let component;
    let instance;

    beforeEach(() => {
        MockedComponent = jest.fn();
        MenuHoc = slidingMenuHOC(MockedComponent);
        didMountSpy = jest.spyOn(MenuHoc.prototype, 'componentDidMount');
        didUnmountSpy = jest.spyOn(MenuHoc.prototype, 'componentWillUnmount');
        jest.spyOn(window, 'addEventListener');
        jest.spyOn(window, 'removeEventListener');
        component = shallow(<MenuHoc />);
        instance = component.instance(); 
    });

    afterEach(() => {
        window.addEventListener.mockRestore();
        window.removeEventListener.mockRestore();
    });
 
    test('Should HOC renders wrapped elem', () => {
        expect(component).toMatchSnapshot();
    });

    describe('HOCMenu cicles', () => {
        test('Did mount method when started', () => {
            expect(didMountSpy).toHaveBeenCalledTimes(1);
            expect(window.addEventListener).toHaveBeenCalledTimes(1);
        });
        test('Unmount method after elem is unmouted', () => {
            expect(didUnmountSpy).toHaveBeenCalledTimes(0);
            component.unmount();
            expect(didUnmountSpy).toHaveBeenCalledTimes(1);
        });
         test('Should remove listener when unmount', () => {
             component.unmount();
             expect(window.removeEventListener).toHaveBeenCalledTimes(1);  

         });
    });
    describe('HOCMenu methods', () => {

         test('Should resizeAlg to be called when peelOfMenu is', () => {
            instance.resizeAlg = jest.fn(); 
            instance.peelOffMenu();
            expect(instance.resizeAlg).toHaveBeenCalledTimes(1); 
        });

        test('Should resizeAlg change state', () => {
            instance.resizeAlg(800);
            expect(component.state().fixed).toBe(true);
        });
    });
  
});