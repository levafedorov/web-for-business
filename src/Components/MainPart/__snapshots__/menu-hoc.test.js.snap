// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOCMenu component Should HOC renders wrapped elem 1`] = `
<mockConstructor
  position={
    Object {
      "position": "relative",
      "top": 0,
    }
  }
/>
`;
